### PR TITLE
Add Kotlin DSL support

### PR DESCRIPTION
**Pull request by martintreurnicht**
Add extensions to add support for Kotlin DSL

### CODE REVIEW

**Comment by REPLicated** [comment]
@etiennestuder : Is there any chance something like this could get merged in the near future?

**Comment by dotCipher** [comment]
Would love to see this merged too, as a fan of both Kotlin(-dsl) and jOOQ

**Comment by etiennestuder** [comment]
I will get to this. But not right now.

**Comment by dotCipher** [comment]
Any updates on this?

File: build.gradle [hunk 1]
```diff
@@ -1,6 +1,8 @@
 plugins {
-    id 'com.gradle.build-scan' version '1.16'
-    id 'nu.studer.plugindev' version '1.0.12'
```
Comment (martintreurnicht) [inline build.gradle]: The latest version of `nu.studer.plugindev` is causing a conflict with one of the kotlin dependencies it seems, it's trying to create a task called `pluginUnderTestMetadata` but it already exists. The easiest solution would be to release a version of `nu.studer.plugindev` that is more tolerant of task names conflicts

**Comment by martintreurnicht** [comment]
@dotCipher If you're not willing to wait for this PR you can always just copy the extensions file that i created and place it in buildSrc of your project, should work just fine.

File: README.md [hunk 2]
```diff
@@ -161,13 +213,75 @@ jooq {
 }
 ```
 
+Kotlin DSL
+```kotlin
+import nu.struder.gradle.jooq.*
```
Comment (aclowkey) [inline README.md]: should be `studer` ?

Comment (martintreurnicht) [reply README.md]: Nice catch, thanks!

Comment (green-s) [reply README.md]: Readme is fixed but the actual package name is still misspelt.

Comment (martintreurnicht) [reply README.md]: Ah, yeah sorry fixed now

**Comment by martintreurnicht** [comment]
@etiennestuder Any update on when we can expect this to be merged?

**Comment by javenwang** [comment]
> @dotCipher If you're not willing to wait for this PR you can always just copy the extensions file that i created and place it in buildSrc of your project, should work just fine.

@martintreurnicht I've copied `JooqExtensionExts.kt` to `buildSrc/src/main/kotlin/nu/studer/gradle/jooq/` but got an error message
```
* What went wrong:
Script compilation error:

  Line 76:     "sample"(sourceSets["main"]) {
               ^ Expression '"sample"' of type 'String' cannot be invoked as a function. The function 'invoke()' is not found
```

anything I missed?

File: README.md [hunk 3]
```diff
@@ -161,13 +213,75 @@ jooq {
...
+jooq {
+  version = "3.11.2"
+  edition = JooqEdition.OSS
+  "sample"(sourceSet["main"]) {
+    jdbc {
+      driver = "org.postgresql.Driver"
+      url = "jdbc:postgresql://localhost:5432/sample"
+      user = "some_user"
+      password = "secret"
+      properties {
```
Comment (javenwang) [inline README.md]:
```suggestion
      properties = listOf(
```
there is no `Jdbc.properties()` fun

**Comment by tomfi** [comment]
@javenwang i dont know if it is still relevant, but for anyone that follow/get to this thread until it will be merged, this is what I did in order to make it work with kotlin dsl:

1. copy `JooqExtensionExts.kt` to `buildSrc/src/main/kotlin` (without the package declaration)
1. Add `implementation("nu.studer:gradle-jooq-plugin:3.0.3")` to `build.gradle.kts` under `buildSrc`:
```
plugins {
  `kotlin-dsl`
}

repositories {
  jcenter()
}

dependencies {
  // TODO: remove when jooq gradle plugin has official kotlin dsl support
  //       https://github.com/etiennestuder/gradle-jooq-plugin/pull/91
  implementation("nu.studer:gradle-jooq-plugin:3.0.3")
}
```
In main `build.gradle.kts` add:
1. `apply(plugin = "nu.studer.jooq")`
1. Under `dependencies` block add:
    1. `implementation("org.jooq:jooq:3.11.11")`
    1. `"jooqRuntime"("org.postgresql:postgresql:42.2.5")` (or any other driver)
1. (Optional) use a custom source-set by adding `sourceSets.create("custom")`
1. Example jooq configuration in kotlin dsl:
```
jooq {
  version = "3.11.11"
  edition = JooqEdition.OSS
  "db"(sourceSets["custom"]) {
    jdbc {
      driver = "org.postgresql.Driver"
      url = "jdbc:postgresql://localhost:5432/postgres"
      user = "postgres"
      password = "postgres"
    }
    generator {
      name = "org.jooq.codegen.DefaultGenerator"
      database {
        name = "org.jooq.meta.postgres.PostgresDatabase"
        includes = ".*"
        excludes = "schema_version"
        inputSchema = "public"
        isIncludeTables = true
        isIncludeRoutines = true
        isIncludePackages = false
        isIncludeUDTs = true
        isIncludeSequences = true
      }
      generate {
        isDeprecated = false
        isRecords = true
        isImmutablePojos = true
        isFluentSetters = true
      }
      target {
        packageName = "io.example.generated"
      }
      strategy {
        name = "org.jooq.codegen.DefaultGeneratorStrategy"
      }
    }
  }
}
```

**Comment by davinkevin** [comment]
Really waiting for this to be merge 👍

**Comment by dotCipher** [comment]
@etiennestuder any traction on your review?

**Comment by zeveras** [comment]
Any news about this PR? It is really important.

**Comment by bolzani** [comment]
This PR is essential to a project I'm working on. Could we maybe get an estimate on when we'd get a merge?

**Comment by marshallpierce** [comment]
As an overstretched OSS maintainer myself, I'm sure Etienne knows about this, but is busy, and will get to it when he can. In the meantime you can always put the jooq gradle config in a groovy gradle file and apply it from your kotlin build as a tolerable workaround.

**Comment by kramerc** [comment]
I managed to get the changes of this PR installed over [JitPack](https://jitpack.io/) but I had to make a fork to make changes to `build.gradle` in order for it to work properly. https://github.com/kramerc/gradle-jooq-plugin/commit/e7a48611041dc39423fe5bea73123f32525d2a1c

To use it over JitPack, add this to `settings.gradle`:
```kotlin
pluginManagement {
    repositories {
        gradlePluginPortal()
        maven(url = "https://jitpack.io")
    }
}
```

Then add the plugin via the fork:
```kotlin
plugins {
    id("com.github.kramerc.gradle-jooq-plugin") version "jitpack-SNAPSHOT"
}
```

**Comment by XhstormR** [comment]
I found the [asciidoctor gradle plugin to use the delegateClosureOf method](https://github.com/asciidoctor/asciidoctor-gradle-plugin#_footnoteref_1) in the Kotlin DSL as a solution.

Can it be used here?

**Comment by kramerc** [comment]
@XhstormR I tried playing around with `delegateClosureOf` to see if I could get this plugin to work without the changes in this PR. I didn't use it in the end but here's what I came up with. It's not as pretty as it could be but it works. I had to add null checks for `jdbc`, `generator`, and `database` otherwise I'd get a `NullPointerException`.

```kotlin
jooq {
    version = "3.11.11"

    val jooqConfig = JooqConfiguration("sample", sourceSets["main"], Configuration())
    jooq.whenConfigAdded.invoke(jooqConfig)

    jooqConfig.configuration.apply {
        jdbc = jdbc ?: Jdbc()
        jdbc.apply {
            driver = "org.postgresql.Driver"
            url = sampleUrl
            user = sampleUsername
            password = samplePassword
        }
        generator = generator ?: Generator()
        generator.apply {
            database = database ?: Database()
            database.apply {
                name = "org.jooq.meta.postgres.PostgresDatabase"
                inputSchema = "public"
            }
        }
    }
}
```

**Comment by XhstormR** [comment]
@kramerc Thank you, I have improved a little on your basis:
```kotlin
jooq {
    version = "3.11.11"

    val jooqConfig = JooqConfiguration("main", sourceSets["main"], Configuration().apply {
        logging = org.jooq.meta.jaxb.Logging.WARN

        jdbc = Jdbc().apply {
            driver = "org.postgresql.Driver"
            url = "jdbc:postgresql://127.0.0.1:5432/postgres"
            user = "123"
            password = "123456"
        }

        generator = Generator().apply {
            database = Database().apply {
                name = "org.jooq.meta.postgres.PostgresDatabase"
                inputSchema = "erp_dev"
            }
            generate = Generate().apply {
                isJavaTimeTypes = true
            }
            target = Target().apply {
                packageName = "com.xhstormr.erp.domain.model.entity"
            }
        }
    })

    whenConfigAdded.invoke(jooqConfig)
}
```

**Comment by martintreurnicht** [comment]
For those interested to use this right now, all you really need is to copy `JooqExtensionExts.kt` into your gradle project's `buildSrc`, then just use it any `build.gradle.kts` file (you should be able to just import the `jooq` extension function.

The reason for the PR was just so that others can benefit from this, but it can be used without it.

**Comment by etiennestuder** [comment]
@martintreurnicht I've come to the conclusion that in the near future, I will not find the time to dutifully review and integrate this PR, and ensure it works across JDKs and Gradle versions, etc. Neither will I have the time to maintain the new feature going forward, and handle the support questions that might follow once this new feature is released.

But, I've made your [proposal](https://github.com/etiennestuder/gradle-jooq-plugin/pull/91#issuecomment-502757938) of achieving Kotlin DSL support via the JooqExtensionExts.kt class in buildSrc a documented example in the README.

The documented example has been [pushed to master](https://github.com/etiennestuder/gradle-jooq-plugin/commit/f1c06ce2cc78a307f2481bd474501d3004a983d5), but I did not have a chance yet to run it and verify that it works since I'm currently on a plane across the ocean, and downloading the 27MB Kotlin compiler will take longer than the remainder of my flight. I will verify the example once I land.

If you could take a look at the example and spot any improvements, that would be highly appreciated.

I would like to thank you for putting the time into this PR, the advice you gave to others, and your patience - and I hope to have your understanding that I cannot make this a first-class feature of the gradle-jooq-plugin at this time.

**Comment by martintreurnicht** [comment]
@etiennestuder Although sad this will not make it into the source base, I completely understand. We all have limited time to contribute to the open source community. My goal was simply to share the work that i've done so that others could benefit from it too.

The README and the inclusion of the ext file i think is really good enough, and looks good to me.

Thanks for spending the time to include this and thanks to everyone for their feedback 🙇

**Comment by ratijas** [comment]
In the meantime we can also enjoy the dirty solution using `withGroovyBuilder` namespace, which enables Kotlin to "invoke" strings like groovy internally does. So basically, inside this scope the example (`"sample"(sourceSets["main"])`) should work as expected, and the error `Expression '"sample"' of type 'String' cannot be invoked as a function. The function 'invoke()' is not found` will be resolved.

Of course, this is just a hacky-ish work-around, but it is the one which is good to know about, as it may as well apply to other non-ported plugins.

**Comment by Zanget87** [comment]
@ratijas could you elaborate or show a working example using the `withGroovybuilder` please? I am also having a hard time trying to integrate this plugin with my kotlin project and I wanted to give a try to this option but I keep getting the error: `Expression '"sample"' of type 'String' cannot be invoked as a function. The function 'invoke()' is not found. `

@martintreurnicht did you get this error during integration of the plugin to an actual project?

Any help will be appreciated. Thanks!

**Comment by etiennestuder** [comment]
@martintreurnicht The gradle-jooq-plugin v10.1 comes with first-class support for the Gradle Kotlin DSL.
https://github.com/etiennestuder/gradle-jooq-plugin/releases/tag/v10.1

It relies on Lukas's `jooq-meta-kotlin` library.
The only thing required in the user's build file is the declaration of `import org.jooq.meta.kotlin.*`

**Pull request closed without merging**